### ISSUE TITLE
Doubt about formulating ensemble / MoE networks

### ISSUE THREAD

**Issue by shyams2** [patrick-kidger/equinox#92]
Hi

Firstly, thanks for this great library! I've been exploring equinox and I really like the design and abstraction. That said, I'm rather new to the library and I had a question about an efficient way to compute over an ensemble of networks. Usually, the way I've been doing this for mixture of experts models is

```
init_e, apply_e = MLP(...)
# assuming that final layer has softmax
init_g, apply_g = MLP([..., num_experts])

# initializing parameters:
Keys = random.split(key = ..., num_experts + 1)
params_e = vmap(init_e)(keys[:-1])
params_g = vmap(init_g)(keys[-1])

# applying for a single example:
experts_out = vmap(apply_e, (0, None))(params_e, input)
gate_out = apply_g(params_g, input)

out = einsum('nd, n -> d', out, gate_out)
```

I'm curious what's the best way to do this in equinox. A naive solution that comes to my mind is sequentially computing the individual experts inside `__call__` with a for loop and then combining with the gating network. As for the initialization, the way I've tried doing it is using an attribute of the class to be a list of MLPs (our experts) which are sequentially initialized in a for loop. Any thoughts?

**Comment by patrick-kidger** [comment]
[`filter_vmap`](https://docs.kidger.site/equinox/api/filtering/filtered-transformations/#equinox.filter_vmap) has an example in which it is used to construct and evaluate an ensemble of models. This is fundamentally equivalent to the sort of thing you're doing above, the "filter" just handles all of the non-array-types appropriately. (Since Equinox takes a more general approach to this kind of thing than most libraries, and allows non-array-types in its PyTrees.)

In particular you don't need to compute things sequentially (which would probably be quite slow).

Here's a complete example following your approach:
```python
import equinox as eqx
import jax
import jax.nn as jnn
import jax.numpy as jnp
import jax.random as jr

zero_or_none = lambda x: 0 if eqx.is_array(x) else None

class MixtureOfMLPs(eqx.Module):
    mlps: eqx.nn.MLP
    gate: eqx.nn.MLP

    def __init__(self, in_size, out_size, width_size, depth, mixture_size, *, key):
        mkey, gkey = jr.split(key)
        mkeys = jr.split(mkey, mixture_size)

        build_mlp = lambda k: eqx.nn.MLP(in_size, out_size, width_size, depth, key=k)
        self.mlps = eqx.filter_vmap(build_mlp, out=zero_or_none)(mkeys)

        self.gate = eqx.nn.MLP(in_size, mixture_size, width_size, depth, key=gkey)

    def __call__(self, x):
        # vmap down the wrapped function
        # don't vmap through the arguments
        vals = eqx.filter_vmap(self.mlps, fn=zero_or_none, args=(None,))(x)
        gate = self.gate(x)
        gate = jnn.softmax(gate)
        return jnp.einsum("nd, n -> d", vals, gate)

in_size = 2
out_size = 3
width_size = 4
depth = 5
mixture_size = 6
key = jr.PRNGKey(0)
mkey, dkey = jr.split(key)

model = MixtureOfMLPs(in_size, out_size, width_size, depth, mixture_size, key=mkey)
data = jr.normal(dkey, (in_size,))
assert model(data).shape == (out_size,)
```

Here you can see how we really take advantage of the "everything is a PyTree" notion by vmap'ing class constructors and such.

**Comment by shyams2** [comment]
Perfect! Just one detail I don't think I can quite understand fully:  the purpose of the `zero_or_none` function. The way I understand it, that's the function that filters? (decides whether we should be vmapping at the leaf level or not)

**Comment by patrick-kidger** [comment]
Yep, that's correct. zero-or-none is actually already the default behaviour for inputs, but for outputs the default is just "always zero" (and error if you get something that's not an array). For the output of `filter_vmap`, using a filter function (and not just a fixed integer/`None`) is technically experimental behaviour so it's not enabled by default.

**Comment by shyams2** [comment]
I see! Thanks a lot for the clarification
